### ISSUE TITLE
libATen.so is too large.

### ISSUE THREAD

**Issue by junjihashimoto** [hasktorch/hasktorch#134]
The size of libATen.so is 1.4GB. It is too large.
Could you reduce the size?

```
root@bbbbde230ac4:/hasktorch# git log -1
commit d2cd93cac56ed8d5194ebf214ada810731d2e34e
Merge: 3afed12 512b385
Author: austinvhuang <austinh@alum.mit.edu>
Date:   Wed Dec 19 18:13:26 2018 -0500

    Merge branch 'master' of https://github.com/hasktorch/hasktorch
root@bbbbde230ac4:/hasktorch# ls -hl  ffi/deps/aten/build/lib/
total 1.4G
lrwxrwxrwx 1 root root   12 Dec 21 18:13 libATen.so -> libATen.so.1
-rw-r--r-- 1 root root 1.4G Dec 22 02:11 libATen.so.1
```

**Comment by austinvhuang** [comment]
Hi @junjihashimoto, hmm that's not normal, on my local OSX laptop as well as the circleCI linux server build and in both it's ~ 12-13 MB.

What steps are you using to build? The `build-aten.sh` script? Can you share some logs here?

**Comment by junjihashimoto** [comment]
Hi @austinvhuang,
Thank you for your quick response.

I used ```build-aten.sh``` in this [Dockerfile](https://github.com/junjihashimoto/hasktorch-docker/blob/master/Dockerfile).
The docker-image is the latest one of https://cloud.docker.com/repository/docker/junjih/hasktorch.
The environment of building the image is nvidia/cuda:9.0-cudnn7-devel-ubuntu16.04.
The log file is gone, now I regenerate the image to make a building logs.

My email is junji.hashimoto@gmail.com. (I want to send you an email, but I can not find your email.)

Best Regards,
Junji Hashimoto

**Comment by junjihashimoto** [comment]
You can see the configuration of ATen in current log-file, though the building a image does not finish, 
https://gist.github.com/junjihashimoto/90ebabfa59c2a9a2d03ceea53887c33d

**Comment by junjihashimoto** [comment]
The complete log is here.
https://gist.git.gree-dev.net/junji-hashimoto/bb1ced5248085613d431ebc80204dcfe

**Comment by austinvhuang** [comment]
Posting follow-up from the slack discussion.

I created a CPU-only variant of your dockerfile, it’s contents are:

```FROM ubuntu:18.04
MAINTAINER Junji Hashimoto "junji.hashimoto@gree.net"
ENV LANG en_US.UTF-8
ENV LC_ALL en_US.UTF-8
RUN apt-get update -qq && apt-get -y --force-yes install software-properties-common apt-transport-https zlib1g-dev
RUN add-apt-repository ppa:hvr/ghc
RUN apt-get update -qq && apt-get -y --force-yes install curl  git build-essential libtinfo-dev libssl-dev  ghc-8.4.4 cabal-install-head
ENV PATH /opt/ghc/bin:$PATH
RUN cabal update
RUN git clone --recursive https://github.com/hasktorch/hasktorch.git
RUN apt-get install -y --force-yes cmake python3-pip python3-yaml
RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 1
RUN cd hasktorch/ffi/deps && ./build-aten.sh && cd ../../..
RUN cd hasktorch && make init && cd ..
RUN cd hasktorch && cabal new-build all && cd ..```

I confirmed that the CPU-only version is tiny:
```-rw-r--r-- 1 root root  13M Dec 26 17:18 libATen.so.1
lrwxrwxrwx 1 root root   12 Dec 26 17:18 libATen.so -> libATen.so.1```

I reproduced the CUDA build and confirmed it’s large:
```lrwxrwxrwx 1 root root   12 Dec 26 14:22 libATen.so -> libATen.so.1
-rw-r--r-- 1 root root 1.4G Dec 26 14:23 libATen.so.1```

Therefore the diff really is cuda vs. non-cuda library builds.

You can check the detailed contents with `nm -Ds libATen.so.1` and see the symbols:

```root@2db98fbf07f7:/hasktorch/ffi/deps/aten/build/lib# nm -Ds libATen.so.1  | head -20
0000000000e13178 B .gomp_critical_user_blasgemm
                 U GOMP_critical_end
                 U GOMP_critical_name_end
                 U GOMP_critical_name_start
                 U GOMP_critical_start
                 U GOMP_parallel
00000000002e1a57 T THAlloc
0000000000952707 T THAtomicAdd
000000000095286e T THAtomicAddLong
000000000095299b T THAtomicAddPtrdiff
...```

This confirms the cuda build has a much larger symbol table compared to the cpu build:
cpu:
```root@2db98fbf07f7:/hasktorch/ffi/deps/aten/build/lib# nm -Ds libATen.so.1  | wc -l   
10885```
cuda:
```root@7f220cbafeb9:/hasktorch/ffi/deps/aten/build/lib# nm -Ds libATen.so.1 | wc -l   
77262```

so bottom line as of today it seems the options are to choose one: compact cpu-bound shared library build, or fat cuda-enabled shared library build.

We can look into the pytorch distribution and contributors to see if they have any tricks to shave off what gets pulled into the shared library.

In the future I’d like to do something similar to pytorch’s distribution model where the shared library is pre-built because the build process is a big impediment to people getting started. If the story is the same w/ the pytorch distribution though, a large library may just be a limitation of cuda integration, but at least people won’t have to build it themselves.
